lib/Optimizer/Lower/Lower.cpp

```diff
@@ -1242,13 +1243,51 @@ static void lowerChannelShuffleNode(Function *F, CompilationContext &cctx,
   replaceAllUsesOfWith(cctx.loweredInfoMap, CSN.getResult(), R2);
 }
 
+// Lower Reduce w/ multiple axes to a single axis, by transposing the input
+// so all the axes to reduce are consecutive, the reshape axes into a single
+// one, to finally perform reduce on a single axis.
+template <typename RnTy>
+Node *reduceMultiAxesOpsHelper(Function *F, CompilationContext &cctx,
+                               RnTy &RN) {
+  auto in = RN.getBatch();
+  auto axes = RN.getAxes();
+  auto name = RN.getName().str();
+
+  std::vector<unsigned_t> perm(in.dims().size());
+  std::vector<unsigned_t> permOld(in.dims().size());
+  std::iota(std::begin(perm), std::end(perm), 0);
+  std::iota(std::begin(permOld), std::end(permOld), 0);
+  for (dim_t i = 1; i < axes.size(); i++) {
+    perm.erase(perm.begin() + axes[i]);
+    perm.insert(perm.begin() + axes[0] + i, axes[i]);
+  }
+
+  Node *out = in.getNode();
+  // Transpose so that axes to remove are consecutive.
+  if (permOld != perm) {
+    out = F->createTranspose(name + "_tr_in", in, perm);
```

Comment:
Hm, this works, though I'd be curious in what cases this gets you better perf than simply chaining multiple single-axis reductions together. With this transpose solution you have one fewer kernel launch and perhaps the backend can optimize it better. With the other version you don't have to transpose the input tensor beforehand.

Reply:
Actually you are right here. I cannot review the change against our kernels at the moment but for CPU, transpose should slow down things. Quickly tried both with CPU and various large reduce ops, and indeed your proposal is faster. Reworked...